### PR TITLE
Refactoring poutines into Messengers and handlers

### PR DESCRIPTION
**Pull request by adscib**
Refactors poutines into messengers that only transform messages and poutines that wrap a computation and contain a messenger.

### CODE REVIEW

**Comment by eb8680** [comment]
Some tests related to enumeration and `IndepPoutine` are failing.  I'll investigate.

**Comment by adscib** [comment]
Looks like this was caused by some unnecessary changes I made to the code. Fixed now.

**Comment by eb8680** [comment]
> Looks like this was caused by some unnecessary changes I made to the code. Fixed now.

I thought the behavior you removed (reverting the scale and independence contexts in the second pass) was what we wanted?

**Comment by adscib** [comment]
I think so, but it doesn't matter for now. I would recommend to leave it as is for now and sort it out later as a part of a more thorough effort to systematise how messengers behave.

**Comment by adscib** [comment]
Sweet! Let me know if you want to switch to using greenlets at some point, I'd be happy to code it.

File: pyro/poutine/escape_poutine.py [hunk 1]
```diff
@@ -32,6 +31,27 @@ def _pyro_sample(self, msg):
         """
         if self.escape_fn(msg):
             msg["done"] = True
-            raise NonlocalExit(msg)
-        else:
-            return super(EscapePoutine, self)._pyro_sample(msg)
+            msg["stop"] = True
+
+            def cont(m):
+                # for enumeration, we need a value at the site to get the shape
```
Comment (fritzo) [inline pyro/poutine/escape_poutine.py]: Why not just call `.shape()`?

Comment (eb8680) [reply pyro/poutine/escape_poutine.py]: I wasn't sure if that was correct, but if so I can change it

File: pyro/poutine/indep_poutine.py [hunk 2]
```diff
@@ -2,40 +2,34 @@
...
-    def _prepare_site(self, msg):
-        """
-        Construct the message that is consumed by ``TracePoutine``;
-        ``cond_indep_stack`` encodes the nested sequence of ``irange`` branches
-        that the site at name is within.
-        """
-        msg["cond_indep_stack"].append(CondIndepStackFrame(self.name, self.counter, self.vectorized))
-        return msg
+    def _process_message(self, msg):
+        msg["cond_indep_stack"].insert(0, CondIndepStackFrame(self.name, self.counter, self.vectorized))
```
Comment (fritzo) [inline pyro/poutine/indep_poutine.py]: This erroneously reverses the direction of `cond_indep_stack`. It worries me that this bug was not caught by tests.

Comment (fritzo) [reply pyro/poutine/indep_poutine.py]: cc @martinjankowiak

**Comment by eb8680** [comment]
Can you write a test that proves that? I'm pretty sure it reverses both the
order of traversal of the poutine stack and the order of insertion into the
independence stack, so that the independence stack ordering is preserved.

File: pyro/poutine/indep_poutine.py (hunk 2 above)
Comment (fritzo) [reply pyro/poutine/indep_poutine.py]: > it reverses both the order of traversal of the poutine stack and the order of insertion into the
independence stack

Oh, that makes sense, thanks for explaining.

Comment (adscib) [reply pyro/poutine/indep_poutine.py]: I actually first just reversed the order of the poutine stack traversal and then your tests caught the problem with the independence stack order so we reversed that too.

**Pull request merged**